## **Part 1: Build a classification model using text data**

**Import the text data, vectorize the review column into an X matrix.  Then run at least three models and select a single best model.  Note that you can also create three models that simply use different types of explanatory variables such as a logistic regression with different n grams or different tokenizers.  Be sure to explain your choice and evaluate this model using the test set.**

In [87]:
import pandas as pd
import numpy as np

In [88]:
url = 'https://raw.githubusercontent.com/freego57/Machine-Learning-Class/master/HW4_Text_data.csv'
text = pd.read_csv(url)
text.head(5)

Review  Recommended
0  "go to" please offer a maxi for mature custome...            1
1  "go to" tee for sz 10+ women.Women who want so...            1
2  "long and warm".These leg warmers are perfect ...            1
3  "tweed" dress.I bought this dress for my siste...            1
4  #1 stunna.I am in lust with this fabulous dres...            1

**Model 1 CountVectorizer with all default parameters**

In [89]:
from sklearn.feature_extraction.text import CountVectorizer
text_train = text['Review']
vect = CountVectorizer().fit(text_train)
X = vect.transform(text_train)
y=text['Recommended']
print(X.shape)
print(y.shape)

(16985, 13010)
(16985,)


In [90]:
# Set up training and test data
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

print(X_train.shape)
print(y_train.shape)

(12738, 13010)
(12738,)


In [91]:
#train the model
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

param_grid = {'C': [0.001, 0.01, 0.1, 1, 10]}
grid = GridSearchCV(LogisticRegression(), param_grid, cv=5,scoring='f1',n_jobs=-1)
grid.fit(X_train, y_train)
print("Best cross-validation score: {:.2f}".format(grid.best_score_))
print("Best parameters: ", grid.best_params_)

Best cross-validation score: 0.94
Best parameters:  {'C': 0.1}


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [92]:
#cross validation on test data
from sklearn.model_selection import cross_val_score
scores = cross_val_score(LogisticRegression(C=0.1), X_test, y_test, cv=5,scoring='f1',n_jobs=-1)
print("Mean cross-validation f1 score: {:.3f}".format(np.mean(scores)))

Mean cross-validation f1 score: 0.934


**Model 2 CountVectorizer with min_df=10 and n-gram=(1,2)**

In [93]:
from sklearn.feature_extraction.text import CountVectorizer
text_train = text['Review']
vect = CountVectorizer(min_df=10,ngram_range=(1,2)).fit(text_train)
X = vect.transform(text_train)
y=text['Recommended']
print(X.shape)
print(y.shape)

(16985, 15906)
(16985,)


In [95]:
# Set up training and test data 
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

print(X_train.shape)
print(y_train.shape)

(12738, 15906)
(12738,)


In [96]:
#train the model
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

param_grid = {'C': [0.001, 0.01, 0.1, 1, 10]}
grid = GridSearchCV(LogisticRegression(), param_grid, cv=5,scoring='f1',n_jobs=-1)
grid.fit(X_train, y_train)
print("Best cross-validation score: {:.2f}".format(grid.best_score_))
print("Best parameters: ", grid.best_params_)

Best cross-validation score: 0.94
Best parameters:  {'C': 0.1}


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [97]:
#cross validation on test data
from sklearn.model_selection import cross_val_score

scores = cross_val_score(LogisticRegression(C=0.1), X_test, y_test, cv=5,scoring='f1',n_jobs=-1)
print("Mean cross-validation f1 score: {:.3f}".format(np.mean(scores)))

Mean cross-validation f1 score: 0.936


**Model 3 TfidfVectorizer with min_df=5**

In [98]:
from sklearn.feature_extraction.text import TfidfVectorizer
text_train = text['Review']
vect = TfidfVectorizer(min_df=5).fit(text_train)
X = vect.transform(text_train)
y=text['Recommended']
print(X.shape)
print(y.shape)

(16985, 4396)
(16985,)


In [99]:
# Set up training and test data
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

print(X_train.shape)
print(y_train.shape)

(12738, 4396)
(12738,)


In [100]:
#train the model
from sklearn.model_selection import GridSearchCV
from sklearn.linear_model import LogisticRegression

param_grid = {'C': [0.001, 0.01, 0.1, 1, 10]}
grid = GridSearchCV(LogisticRegression(), param_grid, cv=5,scoring='f1',n_jobs=-1)
grid.fit(X_train, y_train)
print("Best cross-validation score: {:.2f}".format(grid.best_score_))
print("Best parameters: ", grid.best_params_)


Best cross-validation score: 0.94
Best parameters:  {'C': 10}


/usr/local/lib/python3.6/dist-packages/sklearn/linear_model/_logistic.py:940: ConvergenceWarning: lbfgs failed to converge (status=1):
STOP: TOTAL NO. of ITERATIONS REACHED LIMIT.

Increase the number of iterations (max_iter) or scale the data as shown in:
    https://scikit-learn.org/stable/modules/preprocessing.html
Please also refer to the documentation for alternative solver options:
    https://scikit-learn.org/stable/modules/linear_model.html#logistic-regression
  extra_warning_msg=_LOGISTIC_SOLVER_CONVERGENCE_MSG)


In [101]:
#cross validation on test data
from sklearn.model_selection import cross_val_score
from sklearn.linear_model import LogisticRegression

scores = cross_val_score(LogisticRegression(C=10), X_test, y_test, cv=5,scoring='f1',n_jobs=-1)
print("Mean cross-validation f1 score: {:.3f}".format(np.mean(scores)))

Mean cross-validation f1 score: 0.935


**The best model of the three is model 2--CountVectorizer with min_df=10 and n-gram=(1,2), it has the highest f1 score on test dataset, which shows that it can provide the best prediction for this imbalanced dataset.**

## **Part 2: Build a predictive neural network using Keras**

**Run a multilayer perceptron (feed forward neural network) with two hidden layers on the iris dataset using the keras Sequential interface.Include code for selecting the number of hidden units using GridSearchCV and evaluation on a test-set.  Describe the differences in the predictive accuracy of models with different numbers of hidden units.  Describe the predictive strength of your best model.  Be sure to explain your choice and evaluate this model using the test set.**

In [70]:
iris=pd.read_csv('http://vincentarelbundock.github.io/Rdatasets/csv/datasets/iris.csv')
iris=iris.drop(['Unnamed: 0'], axis=1)
print(iris.Species.nunique())  #3 species
print(iris.head(5))

3
   Sepal.Length  Sepal.Width  Petal.Length  Petal.Width Species
0           5.1          3.5           1.4          0.2  setosa
1           4.9          3.0           1.4          0.2  setosa
2           4.7          3.2           1.3          0.2  setosa
3           4.6          3.1           1.5          0.2  setosa
4           5.0          3.6           1.4          0.2  setosa


In [71]:
X = iris.loc[:, iris.columns != 'Species']
y = iris['Species']

In [72]:
from sklearn.preprocessing import LabelEncoder
labelencoder_y= LabelEncoder()
y = labelencoder_y.fit_transform(y)  #label y's each category to 0,1,2

from keras.utils import to_categorical
y=to_categorical(y,num_classes=3,dtype=int)  #to an array with 3 columns with 0/1 encoded


In [73]:
#split to train and test data and scale the data
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(X,y,random_state=42)
from sklearn.preprocessing import StandardScaler
X_train=StandardScaler().fit_transform(X_train)
X_test=StandardScaler().fit_transform(X_test)

In [74]:
from sklearn.model_selection import GridSearchCV
from keras.wrappers.scikit_learn import KerasClassifier
from keras.models import Sequential
from keras.layers import Dense, Activation

# Function to create model, required for KerasClassifier
def create_model(hiddennodes):
	# create model
	model = Sequential()
	model.add(Dense(hiddennodes, input_dim=4, activation='relu'))
	model.add(Dense(hiddennodes, activation='relu'))
	model.add(Dense(3, activation='softmax'))
	# Compile model
	model.compile(loss='categorical_crossentropy', optimizer='sgd', metrics=['accuracy'])
	return model

model = KerasClassifier(build_fn=create_model, epochs=100, verbose=0)

In [77]:
#using GridSearchCV to find the best hiddennodes numbers
hiddennodes = [5,10,15,20,25,30] 
param_grid = dict(hiddennodes=hiddennodes)
grid = GridSearchCV(estimator=model, param_grid=param_grid)
grid_result = grid.fit(X_train, y_train)
print("Best: %f using %s" % (grid_result.best_score_, grid_result.best_params_))
means = grid_result.cv_results_['mean_test_score']
stds = grid_result.cv_results_['std_test_score']
params = grid_result.cv_results_['params']
for mean, stdev, param in zip(means, stds, params):
    print("%f (%f) with: %r" % (mean, stdev, param))

Best: 0.883399 using {'hiddennodes': 25}
0.765613 (0.124657) with: {'hiddennodes': 5}
0.865217 (0.069462) with: {'hiddennodes': 10}
0.828854 (0.079422) with: {'hiddennodes': 15}
0.855731 (0.097204) with: {'hiddennodes': 20}
0.883399 (0.061817) with: {'hiddennodes': 25}
0.874704 (0.052072) with: {'hiddennodes': 30}


***The best model is with 25 hiddennodes in each layer, with the highest accuracy of 0.883, other models all have lower average accuracy ranging from 0.766 to 0.875. ***

In [80]:
#model with 25 nodes per layer
model = Sequential()
model.add(Dense(25, input_dim=4, activation='relu'))
model.add(Dense(25, activation='relu'))
model.add(Dense(3, activation='softmax'))
# Compile model
model.compile(loss='categorical_crossentropy', optimizer='sgd', metrics=['accuracy'])
model.fit(X_train, y_train, batch_size = 30, epochs = 100)

#evaluate prediction error on test data
print("test-set score: {:.3f}".format(grid.score(X_test, y_test)))

Epoch 1/100
112/112 [==============================] - 0s 420us/step - loss: 1.1845 - accuracy: 0.2768
Epoch 2/100
112/112 [==============================] - 0s 51us/step - loss: 1.1502 - accuracy: 0.3214
Epoch 3/100
112/112 [==============================] - 0s 48us/step - loss: 1.1188 - accuracy: 0.3571
Epoch 4/100
112/112 [==============================] - 0s 49us/step - loss: 1.0895 - accuracy: 0.3750
Epoch 5/100
112/112 [==============================] - 0s 56us/step - loss: 1.0622 - accuracy: 0.4554
Epoch 6/100
112/112 [==============================] - 0s 52us/step - loss: 1.0369 - accuracy: 0.4911
Epoch 7/100
112/112 [==============================] - 0s 62us/step - loss: 1.0127 - accuracy: 0.5982
Epoch 8/100
112/112 [==============================] - 0s 69us/step - loss: 0.9887 - accuracy: 0.6964
Epoch 9/100
112/112 [==============================] - 0s 77us/step - loss: 0.9664 - accuracy: 0.7143
Epoch 10/100
112/112 [==============================] - 0s 63us/step - loss: 0.94

**The model on the test dataset has an accuracy score of 0.895.**